### PR TITLE
Raspberry pi only, No navio or pxfmini needed, rc in and pwm output,without shield

### PR DESCRIPTION
**Pull request by crossa**
Both 8 channels PPM encoder and 8 channels revicer are required.
Before launch px4, ppmdeocde program should be launched.
To download ppmdecode program,
visit https://github.com/crossa/raspberry-pi-px4firmware.
PPM signal can be decoded by GPIO.7

Pxfmini and navio are not popular autopilot hardware in china,
I can handly to purchase it.
So that I use raspberry pi to build autopilot separately.
This dirver help to read the pwm value from shared memory and pushlish it to orb system

To see the video of testing，visit http://player.youku.com/player.php/sid/XMjUxMzczNTEzMg==/v.swf

### CODE REVIEW

**Comment by dagar** [comment]
@bkueng are you a position to review this? Who maintains raspi?

**Comment by crossa** [comment]
@dagar @bkueng  
I  have tested this driver and finished first flight.

Click the follow link to see the video

http://v.youku.com/v_show/id_XMjY5NTQ0OTczNg==.html?spm=a2hzp.8244740.userfeed.5!2~5~5~5!2~A

There is still a pwm output driver I wrote  not be released.

I'v made a new branch , see here  https://github.com/crossa/Firmware/tree/v1.5.5-rpi

If using this branch ,you can make the raspberry pi as the pilot without naive or pxfmini

click the following link to see the information
 http://bbs.5imx.com/forum.php?mod=viewthread&tid=1252361&extra=page%3D1



I'll maintains these driver.

**Comment by crossa** [comment]
@dagar @bkueng  I had released all of the code and config file a moment ago

**Comment by bkueng** [comment]
> Who maintains raspi?

That would be me when I find the time :)

Some questions/feedback based on a quick glance:
- There is some LGPL Licensed code (https://github.com/PX4/Firmware/pull/6619/files#diff-557b726d5a7458996a5e9698d44ca57eR12) and we cannot include that here. I'm afraid that you need to rewrite this driver yourself completely.
- Can you convert all Chinese to English?
- Is the separate cmake config & toolchain really needed? Why not use the existing files?
- is this the ppmdecode program: https://github.com/crossa/raspberry-pi-ppm-rc-in ?
 Any reason why it needs to be a separate binary?

**Comment by crossa** [comment]
@bkueng 
1.  I can rewrite the pca9685 driver if neccery.

2.  I can convert  Chinese to English ,but  I'll keep all the comment written by Chinese.

3.  The  separate cmake config & toolchain is needed to support different prefixes of toolchain . Our engineers often make toolchain  themself，so there are different prefixes. 

4.  This ppmdecode programe  is dependent on pigpio so it was a separate  binary, if you want SBUS via uart  as  input device，you can write another progame to replace the ppmdecode binary and no need to adjust the px4 driver.

I refered to the  rc input of navio, The pwm signal  was bed written into the device dependent on kernel driver and the signal will be read from the device  by px4 driver every certain time, I use the program to instead of the kernel driver otherwise  a kernel driver was needed.  I don't  recommend it beacuse  users can not upgrade the system them self if they are using navio or pxfmini, if kernel has been updated , the rc in kernel module will be missed.  So I use a  separate binary and passing the pwm value by shm and if you want upgrade the linux system then upgrade it，don't worry
 about system failure

**Comment by bkueng** [comment]
> The separate cmake config & toolchain is needed to support different prefixes of toolchain . Our engineers often make toolchain themself，so there are different prefixes.

We should not duplicate files just for that. Related: https://github.com/PX4/Firmware/pull/6528

> This ppmdecode programe is dependent on pigpio so it was a separate binary, if you want SBUS via uart as input device，you can write another progame to replace the ppmdecode binary and no need to adjust the px4 driver.

Thanks for the explanation. That sounds ok. The reason why people do it typically in the kernel is for better timing (reduced latency & jitter).

**Comment by crossa** [comment]
@bkueng When using preemptive process management，run a binary in user space is the best way to make it work at realtime especially in respbian system.
The kernel module is not conducive to real time.


Ok, I'll remove the cake config and change the comment and rewrite the pca9685 userspace driver

**Comment by crossa** [comment]
@bkueng I have rebase the driver and remove the duplicated config.
I have written the comment in English.
I have added a config named posix_rpi_cross_without_navio_or_pxfmini.cmake .
I have  add  __DF_RPI_SINGLE definition  and modify some required driver in the cmake file. 
Two month ago I  fixed some issue for DriverFramework about spi frequency and __DF_RPI_SINGLE definition was required to make it work correctly otherwise the copter will crashed.
The px4_rpi.config is the config file for raspberry pi without navio or pxfmini

Please check the pull request

File: cmake/configs/posix_rpi_cross_without_navio_or_pxfmini [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# 使用这个文件可在上位机交叉编译px4firmware
```
Comment (crossa) [inline cmake/configs/posix_rpi_cross_without_navio_or_pxfmini]: Yes. just one second

File: posix-configs/rpi/px4_rpi.config [hunk 2]
```diff
@@ -0,0 +1,27 @@
+uorb start
```
Comment (crossa) [inline posix-configs/rpi/px4_rpi.config]: Yes,just one second

File: src/drivers/rpi_pca9685_pwm_out/PCA9685.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		usleep(10 * 1000);
+		write_byte(fd, MODE1, oldmode | 0x80);
+
+		close(fd);
+	}
+}
+
+//! PWM a single channel
+/*!
+ \param led channel to set PWM value for
```
Comment (crossa) [inline src/drivers/rpi_pca9685_pwm_out/PCA9685.cpp]: LED channel megs pwm Channel.
In china,pca9685 was known as LED controller.
I'll change these comment

File: src/drivers/rpi_pca9685_pwm_out/PCA9685.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	virtual ~PCA9685();
+	void reset(void);
+	void setPWMFreq(int);
+	void setPWM(uint8_t, int, int);
+	void setPWM(uint8_t, int);
+private:
+	int _i2caddr;
+	int _i2cbus;
+	char busfile[64];
+	uint8_t dataBuffer[BUFFER_SIZE];
```
Comment (crossa) [inline src/drivers/rpi_pca9685_pwm_out/PCA9685.h]: OK

File: src/drivers/rpi_rc_in/rpi_rc_in.cpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+	_data.channel_count = _channels;
+	_data.rssi = 100;
+	_data.rc_lost_frame_count = 0;
+	_data.rc_total_frame_count = 1;
+	_data.rc_ppm_frame_length = 100;
+	_data.rc_failsafe = false;
+	_data.rc_lost = false;
+	_data.input_source = input_rc_s::RC_INPUT_SOURCE_PX4IO_PPM;
+
+	orb_publish(ORB_ID(input_rc), _rcinput_pub, &_data);
```
Comment (crossa) [inline src/drivers/rpi_rc_in/rpi_rc_in.cpp]: One second

File: src/drivers/rpi_rc_in/rpi_rc_in.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+//---------------------------------------------------------------------------------------------------------//
+void RcInput::_measure(void)
+{
+	uint64_t ts;
+	// PWM数据发布
+	// read pwm value from shared memory
+	int i = 0;
+
+	for (i = 0; i < _channels; ++i) {
+		_data.values[i] = (*(this->mem + i) <= 0) ? UINT16_MAX : *(this->mem + i);
```
Comment (crossa) [inline src/drivers/rpi_rc_in/rpi_rc_in.cpp]: I found that if you place the throttle stick on the top of the position,  some motors will stop.
I referred to the driver of navio, they treat the problem as same.
But it's not the issue of the driver,the param of QGC should be changed as following line
1. "PWM DISARMED" and "PWM AUX DISARMED"   should be less than the minimum pwm of your throttle
2. "PWM MAX" and "PWM AUX MAX" should be less than the maxmum pwm of your throttle,
 the maxmum value minus 100 was recommend

File: src/drivers/rpi_pca9685_pwm_out/PCA9685.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+#define LED0_OFF_H 0x9		//LED0 output and brightness control byte 3
+#define LED_MULTIPLYER 4	// For the other 15 channels
+#define ALLLED_ON_L 0xFA    //load all the LEDn_ON registers, byte 0 (turn 0-7 channels on)
+#define ALLLED_ON_H 0xFB	//load all the LEDn_ON registers, byte 1 (turn 8-15 channels on)
+#define ALLLED_OFF_L 0xFC	//load all the LEDn_OFF registers, byte 0 (turn 0-7 channels off)
+#define ALLLED_OFF_H 0xFD	//load all the LEDn_OFF registers, byte 1 (turn 8-15 channels off)
+#define PRE_SCALE 0xFE		//prescaler for output frequency
+#define MAX_PWM_RES 4096        //Resolution 4096=12bit 分辨率，按2的阶乘计算，12bit为4096
+#define CLOCK_FREQ 25000000.0 //25MHz default osc clock
+#define BUFFER_SIZE 0x08  //1 byte buffer
```
Comment (crossa) [inline src/drivers/rpi_pca9685_pwm_out/PCA9685.h]: it is the range of the  dataBuffer

File: src/drivers/rpi_pca9685_pwm_out/PCA9685.cpp [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+ \param fd file descriptor for I/O
+ \param address register address to read from
+ */
+uint8_t PCA9685::read_byte(int fd, uint8_t address)
+{
+
+	return 0;
+
+	uint8_t buff[BUFFER_SIZE];
+	buff[0] = address;
```
Comment (crossa) [inline src/drivers/rpi_pca9685_pwm_out/PCA9685.cpp]: It's not the issue.

**Comment by crossa** [comment]
@bkueng  please check the code.

If there is issue I'll fix it as soon as possible

**Comment by bkueng** [comment]
Thanks for your efforts, @crossa, it's much appreciated. I took your PR and continued it in #7200, there were many small things to improve the code quality, and I thought it would be more efficient if I helped you with that.

Can you please also add documentation how to use the `ppmdecode` program and how to connect the boards, something similar to https://dev.px4.io/en/airframes_multicopter/matrice100.html? You find the repository on https://github.com/PX4/Devguide

**Comment by crossa** [comment]
@bkueng   Here is the readme of the ppmdecode please check it
https://github.com/crossa/raspberry-pi-ppm-rc-in/blob/master/README.md

**Comment by bkueng** [comment]
@crossa are you aware of #7200?

**Comment by crossa** [comment]
@bkueng   I will close this pull request but don't how to close it.
If  I push the code ,the pull request will be reopened.
Please ignore it.

I'm rewriting these code in these days and try to port px4 to generic arm board such  as nano pi.

**Pull request closed without merging**